File: lg-chinese/lexer.js

const { NewToken, TOKEN_TYPE} = require('./token')
module.exports = class NewLexer {
   constructor(code) {
    this.code = code
    this.low = 0
    this.fast = 0
    this.ch = ' '
    this.lexerError = []
   }

   readChar() {
    if (this.fast >= this.code.length) {
      this.ch = 'EOF'
    } else {
      this.ch = this.code[this.fast]
    }
    this.low = this.fast
    this.fast += 1
   }

   nextToken() {
    let token = null;
    this.trimChar()
     switch(this.ch) {
      case 'EOF':
        token = NewToken(TOKEN_TYPE.EOF, this.ch);
        break
      case '设':
        token = NewToken(TOKEN_TYPE.LET, this.ch);
        break
      case '为':
          token = NewToken(TOKEN_TYPE.ASSIGN, this.ch);
          break 
      case '等':
        const ch = this.ch
        if (this.peekChar() === '于') {
          this.readChar()
          token = NewToken(TOKEN_TYPE.ASSIGN, ch + this.ch);
        } else {
          this.pushErrors('关键字 等于 缺少')
        }
        break
      case '乘':
          token = NewToken(TOKEN_TYPE.ASTERISK, this.ch);
        
          break
      case '打':
          const pch = this.ch
          if (this.peekChar() === '印') {
            this.readChar()
            token = NewToken(TOKEN_TYPE.PRINT, pch + this.ch);
          } else {
            this.pushErrors('关键字 打印 缺少')
          }
          break
      case ';':
            token = NewToken(TOKEN_TYPE.SEMICOLON, this.ch);
            break 
      case ',':
            token = NewToken(TOKEN_TYPE.COMMA, this.ch);
            break 
      default:
        if (this.isLetter()) {
          token = NewToken(TOKEN_TYPE.IDENT, this.readLetter())
        } else if (this.isNumber()) {
          token = NewToken(TOKEN_TYPE.INT, this.readNumber())
        } else {
          token = NewToken(TOKEN_TYPE.ILLEGAL, this.ch);
        }
     }
     this.readChar()
     return token
   }
   
   trimChar() {
    while(this.ch === ' ' || this.ch === '\t' || this.ch === '\n') {
      this.readChar()
    }
   }

   isNumber() {
    return /\d/.test(this.ch)
   }

   readNumber() {
    let num = ''
    while(this.isNumber(this.ch) && this.ch != 'EOF') {
      num = `${num}${this.ch}`
      this.readChar()
    }
    if (!this.isNumber(this.ch) && this.ch != 'EOF') {
      this.low -= 1
      this.fast -= 1
    }
    return num
   }


   isLetter() {
    return /[a-zA-Z]/.test(this.ch)
   }
   
   readLetter() {
    let leter = ''
    while(this.isLetter(this.ch) && this.ch != 'EOF') {
      leter = `${leter}${this.ch}`
      this.readChar()
    }
    if (!this.isLetter(this.ch) && this.ch != 'EOF') {
      this.low -= 1
      this.fast -= 1
    }
    return leter
   }

   peekChar() {
    return this.code[this.fast]
   }

   pushErrors(error) {
    this.lexerError.push(error)
   }
}

File: monkey/lexer/lexer.go
package lexer

import (
	"monkey/token"
)

// 生成一个标记token
func newToken(tokenType token.TokenType, ch byte) token.Token {
	return token.Token{Type: tokenType, Literal: string(ch)}
}

type Lexer struct {
	input        string
	position     int  // 当前读取的字符索引
	readPosition int  // 下一个字符索引
	ch           byte // 当前字符值
}

func (l *Lexer) NextToken() token.Token {
	var tok token.Token

	switch l.ch {
	case '=':
		tok = newToken(token.ASSIGN, l.ch)
	case ';':
		tok = newToken(token.SEMICOLON, l.ch)
	case '(':
		tok = newToken(token.LPAREN, l.ch)
	case ')':
		tok = newToken(token.RPAREN, l.ch)
	case ',':
		tok = newToken(token.COMMA, l.ch)
	case '+':
		tok = newToken(token.PLUS, l.ch)
	case '{':
		tok = newToken(token.LBRACE, l.ch)
	case '}':
		tok = newToken(token.RBRACE, l.ch)
	case 0:
		tok.Literal = ""
		tok.Type = token.EOF
	}

	l.readChar()
	return tok
}

func (l *Lexer) readChar() {
	if l.readPosition >= len(l.input) {
		l.ch = 0
	} else {
		l.ch = l.input[l.readPosition]
	}
	l.position = l.readPosition
	l.readPosition += 1
}

// 获取变量名
func (l *Lexer) readIdentifier() string {
	position := l.position
	for isLetter(l.ch) {
		l.readChar()
	}
	// 切片
	return l.input[position:l.position]
}

// 判断是否是一个变量
func isLetter(ch byte) bool {
	return 'a' <= ch && ch <= 'z' || 'A' <= ch && ch <= 'Z' || ch == '_'
}

// 初始化词法分析器
func New(input string) *Lexer {
	l := &Lexer{input: input}
	// 将扫描的字符初始化为0
	l.readChar()
	return l
}
